cmd/file/io.go

```diff
@@ -0,0 +1,235 @@
+		if err != nil {
+			return nil, err
+		}
+	}
+	return nil, nil
+}
+
+// FsStore provides a storage.Putter that writes chunks directly to the filesystem.
+// Each chunk is a separate file, where the hex address of the chunk is the file name.
+type FsStore struct {
+	path string
+}
+
+// NewFsStore creates a new FsStore.
+func NewFsStore(path string) storage.Putter {
+	return &FsStore{
+		path: path,
+	}
+}
+
+// Put implements storage.Putter.
+func (f *FsStore) Put(ctx context.Context, mode storage.ModePut, chs ...swarm.Chunk) (exist []bool, err error) {
+	for _, ch := range chs {
+		chunkPath := filepath.Join(f.path, ch.Address().String())
+		err := ioutil.WriteFile(chunkPath, ch.Data(), 0o666)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return nil, nil
+}
+
+// ApiStore provies a storage.Putter that adds chunks to swarm through the HTTP chunk API.
+type ApiStore struct {
+	Client  *http.Client
+	baseUrl string
+}
+
+// NewApiStore creates a new ApiStore.
+func NewApiStore(host string, port int, ssl bool) putGetter {
+	scheme := "http"
+	if ssl {
+		scheme += "s"
+	}
+	u := &url.URL{
+		Host:   fmt.Sprintf("%s:%d", host, port),
+		Scheme: scheme,
+		Path:   "bzz-chunk",
+	}
+	return &ApiStore{
+		Client:  http.DefaultClient,
+		baseUrl: u.String(),
+	}
+}
+
+// Put implements storage.Putter.
+func (a *ApiStore) Put(ctx context.Context, mode storage.ModePut, chs ...swarm.Chunk) (exist []bool, err error) {
+	for _, ch := range chs {
+		addr := ch.Address().String()
+		buf := bytes.NewReader(ch.Data())
+		url := strings.Join([]string{a.baseUrl, addr}, "/")
+		res, err := a.Client.Post(url, "application/octet-stream", buf)
+		if err != nil {
+			return nil, err
+		}
+		if res.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("upload failed: %v", res.Status)
+		}
+	}
+	return nil, nil
+}
+
+// Get implements storage.Getter.
+func (a *ApiStore) Get(ctx context.Context, mode storage.ModeGet, address swarm.Address) (ch swarm.Chunk, err error) {
+	addressHex := address.String()
+	url := strings.Join([]string{a.baseUrl, addressHex}, "/")
+	res, err := http.DefaultClient.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("chunk %s not found", addressHex)
+	}
+	chunkData, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	ch = swarm.NewChunk(address, chunkData)
+	return ch, nil
+}
+
+// LimitWriteCloser limits the output from the application.
+type LimitWriteCloser struct {
+	io.WriteCloser
+	total int64
+	limit int64
+}
+
+// NewLimitWriteCloser creates a new LimitWriteCloser.
+func NewLimitWriteCloser(w io.WriteCloser, c int64) io.WriteCloser {
+	return &LimitWriteCloser{
+		WriteCloser: w,
+		limit:       c,
+	}
+}
+
+// Write implements io.Writer.
+func (l *LimitWriteCloser) Write(b []byte) (int, error) {
+	if l.total+int64(len(b)) > l.limit {
+		return 0, errors.New("overflow")
+	}
+	c, err := l.WriteCloser.Write(b)
+	l.total += int64(c)
```

Comment:
if err is true, i'm not sure `total` should be incremented, or?

Reply:
from `go doc -all io | grep "type Writer interface -A10`:

```
Write writes len(p) bytes from p to the underlying data stream. It returns
    the number of bytes written from p (0 <= n <= len(p)) and any error
    encountered that caused the write to stop early.
```

in other words, `c` may be > 0 even if `err` is not `nil`.